src/compiler/emitter.ts

```diff
@@ -2580,16 +2595,20 @@ namespace ts {
                 : writeTokenText(token, pos);
         }
 
-        function writeTokenNode(node: Node) {
+        function writeTokenAndCallCallbacks(node: Node, text: string) {

```

Comment:
Maybe we can simplify this by replacing all calls to `writeIfPresent` with `emit`, as `emit` already guards against undefined, will in turn eventually call `writeTokenNode`, and may give us better comment preservation and source-maps.